File: Easy/average-of-levels-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:     # 80ms
    def helper(self, node: Optional[TreeNode], i, sAvg, count):
        if node is None:
            return
        if i < len(sAvg):
            sAvg[i] += node.val
            count[i] += 1
        else:
            sAvg.append(node.val)
            count.append(1)
        self.helper(node.left, i + 1, sAvg, count)
        self.helper(node.right, i + 1, sAvg, count)
    
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        sAvg = []
        count = []
        self.helper(root, 0, sAvg, count)
        return [sAvg[i] / count[i] for i in range(len(sAvg))]

File: Easy/average-of-levels-in-binary-tree.java
import java.util.*;

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution { // 3ms
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> ret = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty()) {
            double sum = 0;
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                sum += node.val;
                if (node.left != null)
                    q.add(node.left);
                if (node.right != null)
                    q.add(node.right);
            }
            ret.add(sum / size);
        }
        return ret;
    }
}